logstash-core/src/main/java/org/logstash/config/ir/compiler/DatasetCompiler.java

```diff
@@ -393,21 +394,18 @@ private static MethodLevelSyntaxElement callFilterFlush(final ClassFields fields
     }
 
     private static MethodLevelSyntaxElement unsetPluginIdForLog4j() {
-        return () -> "org.apache.logging.log4j.ThreadContext.remove(\"plugin.id\")";
-    }
-
-    private static MethodLevelSyntaxElement setPluginIdForLog4j(final AbstractFilterDelegatorExt filterPlugin) {
-        final IRubyObject pluginId = filterPlugin.getId();
-        return generateLog4jContextAssignment(pluginId);
-    }
-
-    private static MethodLevelSyntaxElement setPluginIdForLog4j(final AbstractOutputDelegatorExt outputPlugin) {
-        final IRubyObject pluginId = outputPlugin.getId();
-        return generateLog4jContextAssignment(pluginId);
+        return SyntaxFactory.value("org.apache.logging.log4j.ThreadContext").call(
+                "remove",
+                SyntaxFactory.value("\"plugin.id\"")
+        );
     }
 
-    private static MethodLevelSyntaxElement generateLog4jContextAssignment(IRubyObject pluginId) {
-        return () -> "org.apache.logging.log4j.ThreadContext.put(\"plugin.id\", \"" + pluginId + "\")";
+    private static MethodLevelSyntaxElement setPluginIdForLog4j(final ValueSyntaxElement plugin) {
+        return SyntaxFactory.value("org.apache.logging.log4j.ThreadContext").call(
+                "put",
+                SyntaxFactory.value("\"plugin.id\""),
+                plugin.call("getId").call("toString")
```

Comment:
maybe we should avoid the cost of converting a RubyString to String on every call by introducing a String rawID to the FilterDelegatorExt?

Reply:
The `id` is a `RubyString`

https://github.com/elastic/logstash/blob/master/logstash-core/src/main/java/org/logstash/config/ir/compiler/AbstractFilterDelegatorExt.java#L49

This is certainly a very marginal optimization.